### PR TITLE
Reduce allocations in PSTraceSource.WriteLine(). Part 2

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Significant refactoring `PSTraceSource` class.

## Rationale:

The original code causes significant resource consumption even when tracing is turned off.
Part of the problem was reduced earlier by creating many helper methods. This reduced the allocations when tracing is turned off noticeably, but did not eliminate them all (scope tracing, early argument evaluations).
These problems also limited developers, forcing them to use simple strings instead of displaying really useful information.

## Benefits gained:

1. New C# feature - interpolated string handler - allow us to optimize `PSTraceSource` code:
    - exclude boxing and related allocations
    - reduce cascade allocations
2. Implemented one universal `PSTraceWrite()` method:
    - the method should be used for most types of tracing defined by `PSTraceSourceOptions` enum.
    - all old helper methods (that made a huge number of allocations) were removed. The file size has decreased by 1000 lines, almost twice as much.
    - the method benefits from interpolated string handler that allows to exclude argument boxing and lazily evaluate arguments.
3. Implemented one universal `ref readonly struct PSTraceScope` for scope tracing.
    - the struct is disposable that allows automatically write one message on scope begin and one message on scope end.
    - the struct exclude allocations at trace enabled time.
    - the struct benefits from interpolated string handler that allows to exclude argument boxing and lazily evaluate arguments.
4. `PSTraceSource` class was updated with nullable annotations.
5. All call sites were updated with new code - `PSTraceWrite()` method and `PSTraceScope` struct.

Reviewers can use debugger and the follow command:
```powershell
Trace-Command -PSHost -Name FileSystemProvider -Expression {Get-Item c:\}
```

## PR Context

History #10052.

From https://github.com/PowerShell/PowerShell/pull/10052#pullrequestreview-258505038
> I will look into the `FormatOutputLine` later in a separate PR.

@daxian-dbw The time has come! 😄 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/utils/StructuredTraceSource.cs [hunk 1]
```diff
@@ -916,6 +917,16 @@ internal void WriteLine(string format, object arg1)
             }
         }
 
+        /// <summary> Traces the formatted output when <see cref="PSTraceSourceOptions.WriteLine"/> is enabled.</summary>
+        /// <param name="handler">The interpolated string handler.</param>
+        internal void WriteLine([InterpolatedStringHandlerArgument("")] WriteLineIfInterpolatedStringHandler handler)
+        {
+            if ((_flags & PSTraceSourceOptions.WriteLine) != PSTraceSourceOptions.None)
```
Comment (xtqqczze) [inline src/System.Management.Automation/utils/StructuredTraceSource.cs:924]: Use `Enum.HasFlag`, it is more readable and is now optimised in the Jit.

**Review by xtqqczze — CHANGES_REQUESTED** [review]

**Comment by daxian-dbw** [comment]
This is great! There are more places to apply this, like all the `WriteVerbose` and `WriteWarning` calls in engine. This will save a great deal of string allocations.

**Comment by iSazonov** [comment]
> This is great! There are more places to apply this, like all the `WriteVerbose` and `WriteWarning` calls in engine. This will save a great deal of string allocations.

I suggest doing this in some follow PRs because we need to change dozens of files, which will make it difficult to review this PR.

File: src/System.Management.Automation/utils/StructuredTraceSource.cs [hunk 2]
```diff
@@ -1863,4 +1913,104 @@ protected override string[] GetSupportedAttributes()
...
+
+        /// <summary>Creates an instance of the handler.</summary>
+        /// <param name="literalLength">The number of constant characters outside of interpolation expressions in the interpolated string.</param>
+        /// <param name="formattedCount">The number of interpolation expressions in the interpolated string.</param>
+        /// <param name="trace">Instance of current <see cref="PSTraceSource"/>.</param>
+        /// <param name="shouldAppend">A value indicating whether formatting should proceed.</param>
+        /// <remarks>This is intended to be called only by compiler-generated code. Arguments are not validated as they'd otherwise be for members intended to be used directly.</remarks>
+        public WriteLineIfInterpolatedStringHandler(int literalLength, int formattedCount, PSTraceSource trace, out bool shouldAppend)
+        {
+            if ((trace.Options & PSTraceSourceOptions.WriteLine) != PSTraceSourceOptions.None)
```
Comment (xtqqczze) [inline src/System.Management.Automation/utils/StructuredTraceSource.cs:1933]: See #18386.

```suggestion
            if (trace.Options.HasFlag(PSTraceSourceOptions.WriteLine))
```

**Review by xtqqczze — CHANGES_REQUESTED** [review]

**Comment by daxian-dbw** [comment]
Leveraging the interpolated string handler for `WriteLine` doesn't further reduce allocations for the common cases **where tracing is disabled**, because your previous PR https://github.com/PowerShell/PowerShell/pull/10052 already eliminated allocation (mostly) for `WriteLine` **when tracing is disabled**.

The current change in this PR will reduce more allocations **when tracing is enabled**, but that's not a common scenario and thus not very interesting (reducing allocation is always good, it's just not as important in this case).

Back to the design of the current change, it would require one handler implementation per every logging option, because the handler has built-in assumption of the tracing option to check against. This is obviously not desired. What we want is a single handler implementation that can work for all logging options. To achieve this goal, we need to first do significant refactoring to the existing `PSTraceSource` methods -- changing method like `TraceWarning(xxx)`, `WriteLine(xxx)` and etc. to `Trace(Option option, xxxx)` -- so that each method doesn't have built-in assumption but instead the passed-in `option` instructs what to do with this tracing operation. However, the refactoring work will be massive.

**Comment by iSazonov** [comment]
> Leveraging the interpolated string handler for WriteLine doesn't further reduce allocations for the common cases where tracing is disabled, because your previous PR https://github.com/PowerShell/PowerShell/pull/10052 already eliminated allocation (mostly) for WriteLine when tracing is disabled.

We still do early evaluating of arguments (there is even StringBuilder!). 
When tracing is disabled we reduce allocations even more significantly.
All this already makes this change very useful.

> However, the refactoring work will be massive.

I am trying to keep the PR as small as possible.
My plan:
1. Merge the PR
2. Remove all WriteLine() helpers with new one in follow PR
3. Think whether we want to continue and refactor other trace helpers. 
They are so rarely used that I thought it might not be necessary. But we can unify all methods with one `Trace(Option option, xxxx)` in the next PR if you want.

**Comment by daxian-dbw** [comment]
> When tracing is disabled, we reduce allocations **even more significantly**.
All this already makes this change very useful.

Can you please point me to some examples to prove that?

Also, back to the design of the current change, I have concern about having one handler implementation per every logging option. See the 3rd paragraph in my comment https://github.com/PowerShell/PowerShell/pull/18246#issuecomment-1306052049 for details.

-----------

When `tracing` is disabled, we ideally should never call to the `WriteLine` methods. That's doable for the parameter binders but may be harder for the uses in some other places.

**Comment by iSazonov** [comment]
> > When tracing is disabled, we reduce allocations **even more significantly**.
> > All this already makes this change very useful.
> 
> Can you please point me to some examples to prove the **bold part**?

You can look how `WriteLine(string format, object arg1)` is used (384 times). Most is strings but there are FileInfo, Uri, StringBuilder, ...

> Also, back to the design of the current change, I have concern about having one handler implementation per every logging option. See the 3rd paragraph in my comment https://github.com/PowerShell/PowerShell/pull/18246#issuecomment-1306052049 for details.

Options:
1. Most benefits we get from WriteLine() refactoring. Rest of trace methods we could ignore.
2. Create some handlers since they are very simple.
3. I'd prefer to convert all methods to `Trace(Option option, xxxx)` and use one handler.

**Comment by daxian-dbw** [comment]
> You can look how WriteLine(string format, object arg1) is used (384 times). Most is strings but there are FileInfo, Uri, StringBuilder, ...

Having `FileInfo`, `Uri`, `StringBuilder` passed in to `arg1` in this case doesn't cause any allocation when the tracing is disabled, so it's still not clear to me where the big saving would be from.

> I'd prefer to convert all methods to Trace(Option option, xxxx) and use one handler.

This would be most ideal -- being consistent and modern. That will also allow the pattern to be easily copied in other places. The problem is that the saving in allocation will not be obvious when the tracing is disabled (this is the common scenario), which will make the value/cost of this work less attractive :(

**Comment by daxian-dbw** [comment]
Again, I think when tracing is disabled, we ideally should never call to the `WriteLine` methods. That's doable for the parameter binders but may be harder for the uses in some other places.

**Comment by iSazonov** [comment]
> Again, I think when tracing is disabled, we ideally should never call to the `WriteLine` methods. That's doable for the parameter binders but may be harder for the uses in some other places.

Can you clarify how we could achieve this? 
Since it is in run time we should check `if (_flags.HasFlag(PSTraceSourceOptions.WriteLine))` in every callsite. Do you say we need to move the check out of WriteLine method?

> Having FileInfo, Uri, StringBuilder passed in to arg1 in this case doesn't cause any allocation when the tracing is disabled, so it's still not clear to me where the big saving would be from.

We seem to be in slightly different contexts. :-) I must have made a bad initial description.
Earlier we added many helper methods to remove unnecessary allocations. It was a workaround that didn't eliminate the allocation problem completely.

Now we can remove all these helper methods and use a single method that completely solves the problem. There are still limitations there, but they are not critical since the API is not public.

>> I'd prefer to convert all methods to Trace(Option option, xxxx) and use one handler.

>This would be most ideal -- being consistent and modern. That will also allow the pattern to be easily copied in other places. 

We still can keep using specific names like WriteLine and have _one handler_. For this we should have one more _optional_ parameter in the handler and in the methods. What is pattern from the two ones you like more?

>The problem is that the saving in allocation will not be obvious when the tracing is disabled (this is the common scenario), which will make the value/cost of this work less attractive :(

The benefit is that the new pattern eliminates allocations and errors like `s_tracer.WriteLine("DSC ClassCache: loading file '{0}' added the following classes to the cache: {1}", path, sb.ToString());`.

**Comment by iSazonov** [comment]
> > I'd prefer to convert all methods to Trace(Option option, xxxx) and use one handler.
> 
> This would be most ideal -- being consistent and modern. That will also allow the pattern to be easily copied in other places. The problem is that the saving in allocation will not be obvious when the tracing is disabled (this is the common scenario), which will make the value/cost of this work less attractive :(

Create `Write(PSTraceSourceOptions, interpolatedhandler)` method because `Trace` name is already used.

**Pull request closed without merging**